src/modules/discovery/apiserver.py

```diff
@@ -1,41 +1,102 @@
     """The API server is in charge of all operations on the cluster."""
-    def __init__(self, protocol="https"):
+    def __init__(self):
         Service.__init__(self, name="API Server")
-        self.protocol=protocol
+        
+class MetricsServer(Service, Event):
+    """The Metrics server is in charge of providing resource usage metrics for pods and nodes to the API server."""
+    def __init__(self):
+        Service.__init__(self, name="Metrics Server")
 
 
-# Other devices could have this port open, but we can check to see if it looks like a Kubernetes node
-# A Kubernetes API server will respond with a JSON message that includes a "code" field for the HTTP status code
-@handler.subscribe(OpenPortEvent, predicate=lambda x: x.port==443 or x.port==6443 or x.port==8080)
-class ApiServerDiscovery(Discovery):
-    """API Server Discovery
-    Checks for the existence of a an API Server
+# Other devices could have this port open, but we can check to see if it looks like a Kubernetes api
+# A Kubernetes API service will respond with a JSON message that includes a "code" field for the HTTP status code
+@handler.subscribe(OpenPortEvent, predicate=lambda x: x.port in KNOWN_API_PORTS)
+class ApiServiceDiscovery(Discovery):
+    """API Service Discovery
+    Checks for the existence of K8s API Services
     """
     def __init__(self, event):
         self.event = event
-
+        self.session = requests.Session()
+        self.session.verify = False
+    
     def execute(self):
-        logging.debug("Attempting to discover an API server on {}:{}".format(self.event.host, self.event.port))
-        self.make_request(protocol="https")
-        self.make_request(protocol="http")    
+        logging.debug("Attempting to discover an API service on {}:{}".format(self.event.host, self.event.port))
+        protocols = ["http", "https"]
+        for protocol in protocols:
+            if self.has_api_behaviour(protocol):
+                self.publish_event(K8sApiService(protocol))
 
-    def make_request(self, protocol):
+    def has_api_behaviour(self, protocol):
         try:
-            r = requests.get("{}://{}:{}".format(protocol, self.event.host, self.event.port), verify=False)
-            if ('k8s' in r.text) or ('"code"' in r.text and r.status_code is not 200): 
-                self.event.role = "Master"
-                self.publish_event(ApiServer(protocol=protocol))
+            r = self.session.get("{}://{}:{}".format(protocol, self.event.host, self.event.port))
+            if ('k8s' in r.text) or ('"code"' in r.text and r.status_code is not 200):
+                return True
         except requests.exceptions.SSLError:
             logging.debug("{} protocol not accepted on {}:{}".format(protocol, self.event.host, self.event.port))
         except Exception as e:
             logging.debug("{} on {}:{}".format(e, self.event.host, self.event.port))
 
+
+# Acts as a Filter for services, In the case that we can classify the API,
```

Comment:
I think it might be worth commenting what you are classifying (k8s API server, or metrics server) and how you tell the difference (e.g. we expect k8s API service to respond to a /version end point like this, and the Metrics server to respond like that...)

Reply:
@lizrice  Done